java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java

```diff
@@ -0,0 +1,145 @@
+package com.wavefront.ingester;
+
+import com.google.common.base.Preconditions;
+
+import com.wavefront.common.MetricMangler;
+
+import net.razorvine.pickle.PickleUtils;
+import net.razorvine.pickle.Unpickler;
+import sunnylabs.report.ReportPoint;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Pickle protocol format decoder.
+ * https://docs.python.org/2/library/pickle.html
+ */
+public class PickleProtocolDecoder implements Decoder<byte[]> {
+
+  protected static final Logger logger = Logger.getLogger("agent");
+
+  private final String defaultHostName;
+  private final List<String> customSourceTags;
+  private final MetricMangler metricMangler;
+
+  /**
+   * Constructor.
+   * @param customSourceTags list of tags that should be considered the host.
+   * @param mangler the metric mangler object.
+   */
+  public PickleProtocolDecoder(List<String> customSourceTags,
+                               MetricMangler mangler) {
+    this.defaultHostName = "unknown";
+    Preconditions.checkNotNull(customSourceTags);
+    this.customSourceTags = customSourceTags;
+    this.metricMangler = mangler;
+  }
+
+  /**
+   * Constructor.
+   * @param hostName the default host name.
+   * @param customSourceTags list of source tags for this host.
+   * @param mangler the metric mangler object.
+   */
+  public PickleProtocolDecoder(String hostName, List<String> customSourceTags,
+                               MetricMangler mangler) {
+    Preconditions.checkNotNull(hostName);
+    this.defaultHostName = hostName;
+    Preconditions.checkNotNull(customSourceTags);
+    this.customSourceTags = customSourceTags;
+    this.metricMangler = mangler;
+  }
+
+  @Override
+  public void decodeReportPoints(byte[] msg, List<ReportPoint> out, String customerId) {
+    InputStream is = new ByteArrayInputStream(msg);
+    Unpickler unpickler = new Unpickler();
+    Object dataRaw = null;
+    try {
+      dataRaw = unpickler.load(is);
+      if (!(dataRaw instanceof List)) {
+        throw new IllegalArgumentException("unable to unpickle data");
+      }      
+    } catch (final java.io.IOException ioe) {
+      throw new IllegalArgumentException("unable to unpickle data", ioe);
+    }
+
+    // [(path, (timestamp, value)), ...]
+    List<Object[]> data = (List<Object[]>) dataRaw;
+    for (Object[] o : data) {
+      Object[] details = (Object[])o[1];
+      if (details == null || details.length != 2) {
+        logger.warning("Unexpected pickle protocol input");
```

Comment:
could this happen a lot (and spam logs?) please include some information perhaps at least the port that received the invalid data perhaps